### PR TITLE
EXPLAIN (TYPE IO) support for Aria

### PR DESCRIPTION
**Pull request by jbroll**
Allow Connectors to control column constraints in IOPlanPrinter

* Add a new metadata method . toExplainIOConstraints to the Metadata and ConnectorMetadata interfaces to optionally return only Constraint for use in IOPlanPrinter.
* Interface default method returns the current constraints
* Implement HiveConnector.toExplainIOConstraints() to filter out non-partition key columns.

Are there additional data types that Aria can return in Constraints that can also be partition columns?
* Supported:
   * VarcharType
   * TinyintType, SmallintType, IntegerType, BigintType
   * BooleanType

I need some pointer on how to test this.
```

```

### CODE REVIEW

**Comment by bhhari** [comment]
for the PR desc, you should use the NO RELEASE template

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/IOPlanPrinter.java [hunk 1]
```diff
@@ -546,7 +546,12 @@ private String getVarcharValue(Type type, Object value)
             if (type instanceof BooleanType) {
                 return ((Boolean) value).toString();
             }
-            throw new PrestoException(NOT_SUPPORTED, format("Unsupported data type in EXPLAIN (TYPE IO): %s", type.getDisplayName()));
+
+            // Are there additional types that could be partition columns in Aria that we now need to support?
```
Comment (jbroll) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/IOPlanPrinter.java:550]: Purposefully left this comment to reviews to chime in...

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 2]
```diff
@@ -713,6 +713,17 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
         return ((HiveColumnHandle) columnHandle).getColumnMetadata(typeManager);
     }
 
+    /**
+     * Returns a TupleDomain of constraints that is suitable for ExplainIO
```
Comment (bhhari) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:717]: Explain (Type IO)

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 3]
```diff
@@ -713,6 +713,17 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
...
+     *
+     * Only Hive partition columns that are used in OI planning.
```
Comment (bhhari) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:719]: IO planning instead of OI

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/IOPlanPrinter.java [hunk 4]
```diff
@@ -546,7 +546,12 @@ private String getVarcharValue(Type type, Object value)
             if (type instanceof BooleanType) {
                 return ((Boolean) value).toString();
             }
-            throw new PrestoException(NOT_SUPPORTED, format("Unsupported data type in EXPLAIN (TYPE IO): %s", type.getDisplayName()));
+
```
Comment (bhhari) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/IOPlanPrinter.java:549]: We need to keep the throw exception as we are not changing the existing behavior

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/IOPlanPrinter.java [hunk 5]
```diff
@@ -517,6 +519,7 @@ private FormattedDomain parseDomain(Domain domain)
                     discreteValues -> formattedRanges.addAll(
                             discreteValues.getValues().stream()
                                     .map(value -> getVarcharValue(type, value))
+                                    .filter(Objects::nonNull)
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/IOPlanPrinter.java:522]: what's the purpose of this filter?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/IOPlanPrinter.java [hunk 6]
```diff
@@ -497,7 +499,7 @@ public Void visitTableFinish(TableFinishNode node, IOPlanBuilder context)
             for (Map.Entry<ColumnHandle, Domain> entry : constraint.getDomains().get().entrySet()) {
                 ColumnMetadata columnMetadata = metadata.getColumnMetadata(session, tableHandle, entry.getKey());
                 columnConstraints.add(new ColumnConstraint(
-                        columnMetadata.getName(),
+                                columnMetadata.getName(),
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/IOPlanPrinter.java:502]: wrong indentation

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/IOPlanPrinter.java (hunk 5 above)
Comment (jbroll) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/IOPlanPrinter.java:522]: ohh - Thank you.  @bhhari  requested that .getVarcharValue() should throw as it previously did so this filter is no longer necessary.  I had .getVarcharValue() returning null for unknown types and then dropping those here to provide some type of best effort Explain IO.  I can remove the filter.

**Comment by mbasmanova** [comment]
@jbroll The commit message currently says "[WIP] EXPLAIN (TYPE IO) support for Aria". Would you remove [WIP] and, perhaps, rephrase it to something like "Fix EXPLAIN (TYPE IO) when Aria scan is ON"?
